### PR TITLE
Fix apply_bitmask logit for both CPU and triton versions when shape and stride doesn't match

### PR DESCRIPTION
**Pull request by Jialin**
# Motivation
In vLLM, in order to address [Issue #19493](https://github.com/vllm-project/vllm/issues/19493), [PR #19565](https://github.com/vllm-project/vllm/pull/19565) served as a workaround to mitigate the issue by switching to use torch.compile version instead of triton version.

# Root cause
With some investigation, we found that the error would happen when logits.stride()[0] != logits.shape[-1].

# Changes
- Add a unit test to reproduce the errors
- Fix CPU and Triton-version apply_grammar_bitmask for this scenario

# Tests
We've verified
- New unit tests would failed on trunk
- New unit tests passed with the change (and no new failure, however, there're a lot of failing tests in trunk, 47 failed in tests/python/test_token_bitmask_operations.py)

# Followup
We will follow up with a benchmark comparison to see if we should bring back triton-version or use the new cuda version on vLLM side.

### CODE REVIEW

**Review by Ubospica — APPROVED** [review]
Thank you so much for your contribution!

strides are not considered previously. It's good that this PR fixes this issue.

In the future we will further use nanobind's ArrayList to simplify the logic of passing torch tensor to C++. See https://github.com/mlc-ai/xgrammar/blob/main/cpp/nanobind/nanobind.cc#L45. This can avoid the complex logic of passing shape, stride, and dtype, etc. separately to C++.

**Comment by Jialin** [comment]
@Ubospica The precommit failure is addressed. And also we tried to fix cuda implementation in https://github.com/mlc-ai/xgrammar/pull/394

**Comment by Ubospica** [comment]
@Jialin Thank you so much for your contribution
